ddtrace/profiling/recorder.py

```diff
@@ -1,17 +1,38 @@
 # -*- encoding: utf-8 -*-
 import collections
+import threading
 
-from ddtrace.profiling import _attr
 from ddtrace.vendor import attr
 
 
+class _defaultdictkey(dict):
+    """A variant of defaultdict that calls default_factory with the missing key as argument."""
+
+    def __init__(self, default_factory=None):
+        self.default_factory = default_factory
+
+    def __missing__(self, key):
+        if self.default_factory:
+            v = self[key] = self.default_factory(key)
+            return v
+        raise KeyError(key)
+
+
 @attr.s(slots=True, eq=False)
 class Recorder(object):
     """An object that records program activity."""
 
+    _DEFAULT_MAX_SIZE = 32768
+
+    default_max_size = attr.ib(default=_DEFAULT_MAX_SIZE)
+    """The default maximum number of events."""
+
+    max_sizes = attr.ib(factory=dict)
+    """A dict of {event_type_class: max size} to limit the number of events to record."""
+
     events = attr.ib(init=False, repr=False)
-    max_size = attr.ib(factory=_attr.from_env("DD_PROFILING_MAX_EVENTS", 49152, int))
     event_filters = attr.ib(factory=lambda: collections.defaultdict(list), repr=False)
+    _lock = attr.ib(init=False, repr=False, factory=threading.Lock)
```

Comment:
Was this lock always necessary? Why the addition now?

Reply:
Because the `_defaultdictkey` is not going to be thread-safe… but I think I realize this also has an issue with gevent. I need to rework it :(